File: sdks/go/wdk/multisig/safe.go
// Package multisig provides multi-sig (Safe) smart account support for T402.
package multisig

import (
	"context"
	"crypto/ecdsa"
	"errors"
	"fmt"
	"math/big"

	"github.com/ethereum/go-ethereum"
	"github.com/ethereum/go-ethereum/accounts/abi"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/core/types"
	"github.com/ethereum/go-ethereum/crypto"
	"github.com/ethereum/go-ethereum/ethclient"
)

// SafeClient provides interaction with Safe multi-sig contracts.
type SafeClient struct {
	// Address of the Safe contract
	Address common.Address
	// ChainID for the network
	ChainID *big.Int
	// RPC client
	client *ethclient.Client
	// Cached Safe info
	cachedInfo *SafeInfo
}

// NewSafeClient creates a new Safe client.
func NewSafeClient(config SafeConfig) (*SafeClient, error) {
	if config.Address == (common.Address{}) {
		return nil, errors.New("safe address is required")
	}
	if config.RPCURL == "" {
		return nil, errors.New("RPC URL is required")
	}

	client, err := ethclient.Dial(config.RPCURL)
	if err != nil {
		return nil, fmt.Errorf("failed to connect to RPC: %w", err)
	}

	chainID := config.ChainID
	if chainID == nil {
		chainID, err = client.ChainID(context.Background())
		if err != nil {
			return nil, fmt.Errorf("failed to get chain ID: %w", err)
		}
	}

	return &SafeClient{
		Address: config.Address,
		ChainID: chainID,
		client:  client,
	}, nil
}

// GetInfo retrieves current Safe information (owners, threshold, nonce).
func (s *SafeClient) GetInfo(ctx context.Context) (*SafeInfo, error) {
	owners, err := s.GetOwners(ctx)
	if err != nil {
		return nil, fmt.Errorf("failed to get owners: %w", err)
	}

	threshold, err := s.GetThreshold(ctx)
	if err != nil {
		return nil, fmt.Errorf("failed to get threshold: %w", err)
	}

	nonce, err := s.GetNonce(ctx)
	if err != nil {
		return nil, fmt.Errorf("failed to get nonce: %w", err)
	}

	info := &SafeInfo{
		Address:   s.Address,
		Owners:    owners,
		Threshold: threshold,
		Nonce:     nonce,
		ChainID:   s.ChainID,
	}

	s.cachedInfo = info
	return info, nil
}

// GetOwners returns the list of Safe owners.
func (s *SafeClient) GetOwners(ctx context.Context) ([]common.Address, error) {
	data := GetOwnersSelector
	msg := ethereum.CallMsg{
		To:   &s.Address,
		Data: data,
	}

	result, err := s.client.CallContract(ctx, msg, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to call getOwners: %w", err)
	}

	// Decode the result (address[])
	addressArrayType, _ := abi.NewType("address[]", "", nil)
	args := abi.Arguments{{Type: addressArrayType}}
	decoded, err := args.Unpack(result)
	if err != nil {
		return nil, fmt.Errorf("failed to decode owners: %w", err)
	}

	owners := decoded[0].([]common.Address)
	return owners, nil
}

// GetThreshold returns the required number of signatures.
func (s *SafeClient) GetThreshold(ctx context.Context) (int, error) {
	data := GetThresholdSelector
	msg := ethereum.CallMsg{
		To:   &s.Address,
		Data: data,
	}

	result, err := s.client.CallContract(ctx, msg, nil)
	if err != nil {
		return 0, fmt.Errorf("failed to call getThreshold: %w", err)
	}

	// Decode the result (uint256)
	threshold := new(big.Int).SetBytes(result)
	return int(threshold.Int64()), nil
}

// GetNonce returns the current Safe nonce.
func (s *SafeClient) GetNonce(ctx context.Context) (*big.Int, error) {
	data := NonceSelector
	msg := ethereum.CallMsg{
		To:   &s.Address,
		Data: data,
	}

	result, err := s.client.CallContract(ctx, msg, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to call nonce: %w", err)
	}

	nonce := new(big.Int).SetBytes(result)
	return nonce, nil
}

// IsOwner checks if an address is a Safe owner.
func (s *SafeClient) IsOwner(ctx context.Context, address common.Address) (bool, error) {
	owners, err := s.GetOwners(ctx)
	if err != nil {
		return false, err
	}

	for _, owner := range owners {
		if owner == address {
			return true, nil
		}
	}
	return false, nil
}

// ProposeTransaction creates a new transaction proposal.
func (s *SafeClient) ProposeTransaction(ctx context.Context, tx *SafeTransaction) (*TransactionRequest, error) {
	// Get current nonce if not set
	if tx.Nonce == nil {
		nonce, err := s.GetNonce(ctx)
		if err != nil {
			return nil, fmt.Errorf("failed to get nonce: %w", err)
		}
		tx.Nonce = nonce
	}

	// Calculate transaction hash
	txHash, err := s.GetTransactionHash(ctx, tx)
	if err != nil {
		return nil, fmt.Errorf("failed to get transaction hash: %w", err)
	}

	// Get threshold
	threshold, err := s.GetThreshold(ctx)
	if err != nil {
		return nil, fmt.Errorf("failed to get threshold: %w", err)
	}

	// Create request
	request := &TransactionRequest{
		ID:              generateRequestID(),
		SafeAddress:     s.Address,
		Transaction:     tx,
		TransactionHash: txHash,
		Signatures:      make(map[common.Address]*SafeSignature),
		Threshold:       threshold,
		CreatedAt:       currentTimestamp(),
		ExpiresAt:       currentTimestamp() + DefaultRequestExpirationSeconds,
	}

	return request, nil
}

// GetTransactionHash calculates the Safe transaction hash.
func (s *SafeClient) GetTransactionHash(ctx context.Context, tx *SafeTransaction) (common.Hash, error) {
	// Build calldata for getTransactionHash
	data := buildGetTransactionHashCalldata(tx)
	msg := ethereum.CallMsg{
		To:   &s.Address,
		Data: data,
	}

	result, err := s.client.CallContract(ctx, msg, nil)
	if err != nil {
		return common.Hash{}, fmt.Errorf("failed to call getTransactionHash: %w", err)
	}

	return common.BytesToHash(result), nil
}

// SignTransaction signs a transaction with the given private key.
func (s *SafeClient) SignTransaction(tx *SafeTransaction, privateKey *ecdsa.PrivateKey) (*SafeSignature, error) {
	// Calculate transaction hash
	txHash, err := s.GetTransactionHash(context.Background(), tx)
	if err != nil {
		return nil, fmt.Errorf("failed to get transaction hash: %w", err)
	}

	// Sign the hash
	signature, err := crypto.Sign(txHash.Bytes(), privateKey)
	if err != nil {
		return nil, fmt.Errorf("failed to sign: %w", err)
	}

	// Adjust v value for Safe (add 4 for EOA signature)
	if signature[64] < 27 {
		signature[64] += 27
	}
	signature[64] += 4 // Safe EOA signature type

	signer := crypto.PubkeyToAddress(privateKey.PublicKey)

	return &SafeSignature{
		Signer:        signer,
		Signature:     signature,
		SignatureType: SignatureTypeEOA,
	}, nil
}

// AddSignature adds a signature to a transaction request.
func (s *SafeClient) AddSignature(request *TransactionRequest, sig *SafeSignature) error {
	if request.Signatures == nil {
		request.Signatures = make(map[common.Address]*SafeSignature)
	}

	// Check if signer is an owner
	isOwner, err := s.IsOwner(context.Background(), sig.Signer)
	if err != nil {
		return fmt.Errorf("failed to check owner: %w", err)
	}
	if !isOwner {
		return fmt.Errorf("signer %s is not an owner", sig.Signer.Hex())
	}

	request.Signatures[sig.Signer] = sig
	return nil
}

// ExecuteTransaction executes a Safe transaction with collected signatures.
func (s *SafeClient) ExecuteTransaction(ctx context.Context, request *TransactionRequest, executor *ecdsa.PrivateKey) (*ExecutionResult, error) {
	// Check if ready
	if !request.IsReady() {
		return nil, fmt.Errorf("not enough signatures: have %d, need %d", request.CollectedCount(), request.Threshold)
	}

	// Combine signatures (sorted by signer address)
	packedSigs := s.packSignatures(request.Signatures)

	// Build execTransaction calldata
	calldata := s.buildExecTransactionCalldata(request.Transaction, packedSigs)

	// Send transaction
	executorAddr := crypto.PubkeyToAddress(executor.PublicKey)
	nonce, err := s.client.PendingNonceAt(ctx, executorAddr)
	if err != nil {
		return nil, fmt.Errorf("failed to get executor nonce: %w", err)
	}

	gasPrice, err := s.client.SuggestGasPrice(ctx)
	if err != nil {
		return nil, fmt.Errorf("failed to get gas price: %w", err)
	}

	// Estimate gas
	msg := ethereum.CallMsg{
		From: executorAddr,
		To:   &s.Address,
		Data: calldata,
	}
	gasLimit, err := s.client.EstimateGas(ctx, msg)
	if err != nil {
		return nil, fmt.Errorf("failed to estimate gas: %w", err)
	}

	// Create transaction
	tx := types.NewTransaction(nonce, s.Address, big.NewInt(0), gasLimit, gasPrice, calldata)

	// Sign transaction
	signedTx, err := types.SignTx(tx, types.NewEIP155Signer(s.ChainID), executor)
	if err != nil {
		return nil, fmt.Errorf("failed to sign transaction: %w", err)
	}

	// Send transaction
	err = s.client.SendTransaction(ctx, signedTx)
	if err != nil {
		return nil, fmt.Errorf("failed to send transaction: %w", err)
	}

	return &ExecutionResult{
		TxHash:  signedTx.Hash(),
		Success: true, // Will be updated after receipt
	}, nil
}

// WaitForExecution waits for a transaction to be mined and returns the result.
func (s *SafeClient) WaitForExecution(ctx context.Context, txHash common.Hash) (*ExecutionResult, error) {
	receipt, err := s.waitForReceipt(ctx, txHash)
	if err != nil {
		return nil, err
	}

	return &ExecutionResult{
		TxHash:      txHash,
		Success:     receipt.Status == types.ReceiptStatusSuccessful,
		GasUsed:     receipt.GasUsed,
		BlockNumber: receipt.BlockNumber.Uint64(),
	}, nil
}

// waitForReceipt waits for a transaction receipt.
func (s *SafeClient) waitForReceipt(ctx context.Context, txHash common.Hash) (*types.Receipt, error) {
	for {
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		default:
			receipt, err := s.client.TransactionReceipt(ctx, txHash)
			if err == nil {
				return receipt, nil
			}
			// Keep polling
		}
	}
}

// packSignatures packs signatures sorted by signer address.
func (s *SafeClient) packSignatures(sigs map[common.Address]*SafeSignature) []byte {
	// Sort signers by address
	var signers []common.Address
	for signer := range sigs {
		signers = append(signers, signer)
	}
	sortAddresses(signers)

	// Pack signatures
	var packed []byte
	for _, signer := range signers {
		sig := sigs[signer]
		packed = append(packed, sig.Signature...)
	}

	return packed
}

// buildExecTransactionCalldata builds the execTransaction calldata.
func (s *SafeClient) buildExecTransactionCalldata(tx *SafeTransaction, signatures []byte) []byte {
	// execTransaction(address to, uint256 value, bytes data, uint8 operation, uint256 safeTxGas, uint256 baseGas, uint256 gasPrice, address gasToken, address refundReceiver, bytes signatures)
	data := make([]byte, 0)
	data = append(data, ExecTransactionSelector...)

	// Encode parameters
	addressType, _ := abi.NewType("address", "", nil)
	uint256Type, _ := abi.NewType("uint256", "", nil)
	bytesType, _ := abi.NewType("bytes", "", nil)
	uint8Type, _ := abi.NewType("uint8", "", nil)

	args := abi.Arguments{
		{Type: addressType}, // to
		{Type: uint256Type}, // value
		{Type: bytesType},   // data
		{Type: uint8Type},   // operation
		{Type: uint256Type}, // safeTxGas
		{Type: uint256Type}, // baseGas
		{Type: uint256Type}, // gasPrice
		{Type: addressType}, // gasToken
		{Type: addressType}, // refundReceiver
		{Type: bytesType},   // signatures
	}

	value := tx.Value
	if value == nil {
		value = big.NewInt(0)
	}
	safeTxGas := tx.SafeTxGas
	if safeTxGas == nil {
		safeTxGas = big.NewInt(0)
	}
	baseGas := tx.BaseGas
	if baseGas == nil {
		baseGas = big.NewInt(0)
	}
	gasPrice := tx.GasPrice
	if gasPrice == nil {
		gasPrice = big.NewInt(0)
	}

	encoded, _ := args.Pack(
		tx.To,
		value,
		tx.Data,
		tx.Operation,
		safeTxGas,
		baseGas,
		gasPrice,
		tx.GasToken,
		tx.RefundReceiver,
		signatures,
	)

	data = append(data, encoded...)
	return data
}

// buildGetTransactionHashCalldata builds the getTransactionHash calldata.
func buildGetTransactionHashCalldata(tx *SafeTransaction) []byte {
	data := make([]byte, 0)
	data = append(data, GetTransactionHashSelector...)

	// Encode parameters
	addressType, _ := abi.NewType("address", "", nil)
	uint256Type, _ := abi.NewType("uint256", "", nil)
	bytesType, _ := abi.NewType("bytes", "", nil)
	uint8Type, _ := abi.NewType("uint8", "", nil)

	args := abi.Arguments{
		{Type: addressType}, // to
		{Type: uint256Type}, // value
		{Type: bytesType},   // data
		{Type: uint8Type},   // operation
		{Type: uint256Type}, // safeTxGas
		{Type: uint256Type}, // baseGas
		{Type: uint256Type}, // gasPrice
		{Type: addressType}, // gasToken
		{Type: addressType}, // refundReceiver
		{Type: uint256Type}, // nonce
	}

	value := tx.Value
	if value == nil {
		value = big.NewInt(0)
	}
	safeTxGas := tx.SafeTxGas
	if safeTxGas == nil {
		safeTxGas = big.NewInt(0)
	}
	baseGas := tx.BaseGas
	if baseGas == nil {
		baseGas = big.NewInt(0)
	}
	gasPrice := tx.GasPrice
	if gasPrice == nil {
		gasPrice = big.NewInt(0)
	}
	nonce := tx.Nonce
	if nonce == nil {
		nonce = big.NewInt(0)
	}

	encoded, _ := args.Pack(
		tx.To,
		value,
		tx.Data,
		tx.Operation,
		safeTxGas,
		baseGas,
		gasPrice,
		tx.GasToken,
		tx.RefundReceiver,
		nonce,
	)

	data = append(data, encoded...)
	return data
}

File: sdks/python/t402/src/t402/multisig/safe.py
"""
Safe client implementation for T402 Multi-sig support.
"""

from typing import List, Optional

from eth_account import Account
from web3 import AsyncWeb3, Web3
from web3.types import TxParams

from .constants import (
    GET_OWNERS_SELECTOR,
    GET_THRESHOLD_SELECTOR,
    NONCE_SELECTOR,
    GET_TRANSACTION_HASH_SELECTOR,
    EXEC_TRANSACTION_SELECTOR,
)
from .types import (
    SafeConfig,
    SafeInfo,
    SafeSignature,
    SafeTransaction,
    SignatureType,
    TransactionRequest,
    ExecutionResult,
)
from .utils import generate_request_id, current_timestamp, sort_addresses


class SafeClient:
    """Client for interacting with Safe multi-sig contracts."""

    def __init__(self, config: SafeConfig):
        """
        Initialize SafeClient.

        Args:
            config: Safe configuration with address and RPC URL.
        """
        self.address = Web3.to_checksum_address(config.address)
        self.rpc_url = config.rpc_url
        self._chain_id = config.chain_id
        self._w3: Optional[AsyncWeb3] = None
        self._cached_info: Optional[SafeInfo] = None

    async def _get_w3(self) -> AsyncWeb3:
        """Get or create async Web3 instance."""
        if self._w3 is None:
            self._w3 = AsyncWeb3(AsyncWeb3.AsyncHTTPProvider(self.rpc_url))
        return self._w3

    async def get_chain_id(self) -> int:
        """Get the chain ID."""
        if self._chain_id is None:
            w3 = await self._get_w3()
            self._chain_id = await w3.eth.chain_id
        return self._chain_id

    async def get_info(self) -> SafeInfo:
        """
        Get Safe information (owners, threshold, nonce).

        Returns:
            SafeInfo with current Safe state.
        """
        owners = await self.get_owners()
        threshold = await self.get_threshold()
        nonce = await self.get_nonce()
        chain_id = await self.get_chain_id()

        info = SafeInfo(
            address=self.address,
            owners=owners,
            threshold=threshold,
            nonce=nonce,
            chain_id=chain_id,
        )
        self._cached_info = info
        return info

    async def get_owners(self) -> List[str]:
        """Get list of Safe owners."""
        w3 = await self._get_w3()
        result = await w3.eth.call({"to": self.address, "data": GET_OWNERS_SELECTOR})

        # Decode address[] from ABI
        # Skip offset (32 bytes) and length (32 bytes)
        if len(result) < 64:
            return []

        length = int.from_bytes(result[32:64], "big")
        owners = []
        for i in range(length):
            start = 64 + i * 32
            addr_bytes = result[start + 12 : start + 32]
            owners.append(Web3.to_checksum_address(addr_bytes.hex()))

        return owners

    async def get_threshold(self) -> int:
        """Get the required number of signatures."""
        w3 = await self._get_w3()
        result = await w3.eth.call({"to": self.address, "data": GET_THRESHOLD_SELECTOR})
        return int.from_bytes(result, "big")

    async def get_nonce(self) -> int:
        """Get the current Safe nonce."""
        w3 = await self._get_w3()
        result = await w3.eth.call({"to": self.address, "data": NONCE_SELECTOR})
        return int.from_bytes(result, "big")

    async def is_owner(self, address: str) -> bool:
        """Check if an address is a Safe owner."""
        owners = await self.get_owners()
        address_lower = address.lower()
        return any(o.lower() == address_lower for o in owners)

    async def propose_transaction(
        self, tx: SafeTransaction
    ) -> TransactionRequest:
        """
        Create a new transaction proposal.

        Args:
            tx: The Safe transaction to propose.

        Returns:
            TransactionRequest for collecting signatures.
        """
        # Get nonce if not set
        if tx.nonce is None:
            tx.nonce = await self.get_nonce()

        # Calculate transaction hash
        tx_hash = await self.get_transaction_hash(tx)

        # Get threshold
        threshold = await self.get_threshold()

        # Create request
        from .constants import DEFAULT_REQUEST_EXPIRATION_SECONDS

        now = current_timestamp()
        request = TransactionRequest(
            id=generate_request_id(),
            safe_address=self.address,
            transaction=tx,
            transaction_hash=tx_hash,
            signatures={},
            threshold=threshold,
            created_at=now,
            expires_at=now + DEFAULT_REQUEST_EXPIRATION_SECONDS,
        )

        return request

    async def get_transaction_hash(self, tx: SafeTransaction) -> str:
        """
        Calculate the Safe transaction hash.

        Args:
            tx: The Safe transaction.

        Returns:
            Transaction hash as hex string.
        """
        w3 = await self._get_w3()

        # Build calldata for getTransactionHash
        calldata = self._build_get_transaction_hash_calldata(tx)

        result = await w3.eth.call({"to": self.address, "data": calldata})
        return "0x" + result.hex()

    def sign_transaction(
        self, tx: SafeTransaction, private_key: str
    ) -> SafeSignature:
        """
        Sign a transaction with a private key.

        Args:
            tx: The Safe transaction.
            private_key: Private key hex string.

        Returns:
            SafeSignature containing the signature.
        """
        # Get account from private key
        account = Account.from_key(private_key)

        # We need the transaction hash - this should be calculated beforehand
        # For now, we'll compute it synchronously using Web3
        w3_sync = Web3(Web3.HTTPProvider(self.rpc_url))
        calldata = self._build_get_transaction_hash_calldata(tx)
        result = w3_sync.eth.call({"to": self.address, "data": calldata})
        tx_hash = result

        # Sign the hash
        signed = account.signHash(tx_hash)

        # Adjust v value for Safe (add 4 for EOA signature)
        v = signed.v
        if v < 27:
            v += 27
        v += 4

        # Combine r, s, v
        signature = signed.r.to_bytes(32, "big") + signed.s.to_bytes(32, "big") + bytes([v])

        return SafeSignature(
            signer=account.address,
            signature=signature,
            signature_type=SignatureType.EOA,
        )

    async def sign_transaction_async(
        self, tx: SafeTransaction, private_key: str
    ) -> SafeSignature:
        """
        Sign a transaction asynchronously.

        Args:
            tx: The Safe transaction.
            private_key: Private key hex string.

        Returns:
            SafeSignature containing the signature.
        """
        # Get account from private key
        account = Account.from_key(private_key)

        # Get transaction hash
        tx_hash = await self.get_transaction_hash(tx)
        tx_hash_bytes = bytes.fromhex(tx_hash[2:])

        # Sign the hash
        signed = account.signHash(tx_hash_bytes)

        # Adjust v value for Safe (add 4 for EOA signature)
        v = signed.v
        if v < 27:
            v += 27
        v += 4

        # Combine r, s, v
        signature = signed.r.to_bytes(32, "big") + signed.s.to_bytes(32, "big") + bytes([v])

        return SafeSignature(
            signer=account.address,
            signature=signature,
            signature_type=SignatureType.EOA,
        )

    def add_signature(
        self, request: TransactionRequest, sig: SafeSignature
    ) -> None:
        """
        Add a signature to a transaction request.

        Args:
            request: The transaction request.
            sig: The signature to add.

        Raises:
            ValueError: If signer is not an owner.
        """
        request.signatures[sig.signer.lower()] = sig

    async def execute_transaction(
        self,
        request: TransactionRequest,
        private_key: str,
    ) -> ExecutionResult:
        """
        Execute a Safe transaction with collected signatures.

        Args:
            request: Transaction request with signatures.
            private_key: Private key of the executor.

        Returns:
            ExecutionResult with transaction details.

        Raises:
            ValueError: If not enough signatures.
        """
        if not request.is_ready():
            raise ValueError(
                f"Not enough signatures: have {request.collected_count()}, "
                f"need {request.threshold}"
            )

        w3 = await self._get_w3()
        account = Account.from_key(private_key)

        # Pack signatures sorted by signer address
        packed_sigs = self._pack_signatures(request.signatures)

        # Build execTransaction calldata
        calldata = self._build_exec_transaction_calldata(
            request.transaction, packed_sigs
        )

        # Build transaction
        nonce = await w3.eth.get_transaction_count(account.address)
        gas_price = await w3.eth.gas_price

        # Estimate gas
        gas_estimate = await w3.eth.estimate_gas(
            {
                "from": account.address,
                "to": self.address,
                "data": calldata,
            }
        )

        tx_params: TxParams = {
            "from": account.address,
            "to": self.address,
            "data": calldata,
            "gas": gas_estimate,
            "gasPrice": gas_price,
            "nonce": nonce,
            "chainId": await self.get_chain_id(),
        }

        # Sign and send
        signed_tx = account.sign_transaction(tx_params)
        tx_hash = await w3.eth.send_raw_transaction(signed_tx.rawTransaction)

        return ExecutionResult(
            tx_hash="0x" + tx_hash.hex(),
            success=True,
        )

    async def wait_for_execution(self, tx_hash: str) -> ExecutionResult:
        """
        Wait for a transaction to be mined.

        Args:
            tx_hash: Transaction hash.

        Returns:
            ExecutionResult with final status.
        """
        w3 = await self._get_w3()

        # Wait for receipt
        receipt = await w3.eth.wait_for_transaction_receipt(tx_hash)

        return ExecutionResult(
            tx_hash=tx_hash,
            success=receipt["status"] == 1,
            gas_used=receipt["gasUsed"],
            block_number=receipt["blockNumber"],
        )

    def _pack_signatures(self, signatures: dict) -> bytes:
        """Pack signatures sorted by signer address."""
        # Sort signers by address
        sorted_signers = sort_addresses(list(signatures.keys()))

        # Pack signatures
        packed = b""
        for signer in sorted_signers:
            sig = signatures[signer.lower()]
            packed += sig.signature

        return packed

    def _build_exec_transaction_calldata(
        self, tx: SafeTransaction, signatures: bytes
    ) -> bytes:
        """Build execTransaction calldata."""
        from eth_abi import encode

        # Encode parameters
        encoded = encode(
            [
                "address",
                "uint256",
                "bytes",
                "uint8",
                "uint256",
                "uint256",
                "uint256",
                "address",
                "address",
                "bytes",
            ],
            [
                Web3.to_checksum_address(tx.to),
                tx.value,
                tx.data,
                tx.operation,
                tx.safe_tx_gas,
                tx.base_gas,
                tx.gas_price,
                Web3.to_checksum_address(tx.gas_token),
                Web3.to_checksum_address(tx.refund_receiver),
                signatures,
            ],
        )

        return EXEC_TRANSACTION_SELECTOR + encoded

    def _build_get_transaction_hash_calldata(self, tx: SafeTransaction) -> bytes:
        """Build getTransactionHash calldata."""
        from eth_abi import encode

        nonce = tx.nonce if tx.nonce is not None else 0

        # Encode parameters
        encoded = encode(
            [
                "address",
                "uint256",
                "bytes",
                "uint8",
                "uint256",
                "uint256",
                "uint256",
                "address",
                "address",
                "uint256",
            ],
            [
                Web3.to_checksum_address(tx.to),
                tx.value,
                tx.data,
                tx.operation,
                tx.safe_tx_gas,
                tx.base_gas,
                tx.gas_price,
                Web3.to_checksum_address(tx.gas_token),
                Web3.to_checksum_address(tx.refund_receiver),
                nonce,
            ],
        )

        return GET_TRANSACTION_HASH_SELECTOR + encoded
